src/tidy.zig

```diff
@@ -9,280 +10,516 @@ const Ast = std.zig.Ast;
+            \\hello.txt:1: error: control character code=9
+            \\hello.txt:2: error: control character code=13
+            \\
+        ),
+    );
+}
 
-    if (std.mem.indexOf(u8, source, "== error" ++ ".") != null or
-        std.mem.indexOf(u8, source, "!= error" ++ ".") != null)
-    {
-        return "switch on error to avoid silent anyerror upcast";
+fn tidy_banned(file: SourceFile, errors: *Errors) void {
+    // Vendored code is exempt from bans.
+    if (std.mem.eql(u8, file.path, "src/stdx/vendored/aegis.zig")) return;
+    // Don't ban ourselves!
+    if (std.mem.eql(u8, file.path, "src/tidy.zig")) return;
+
+    const ban_list: []const struct { []const u8, []const u8 } = &.{
+        // Functionality provided by stdx:
+        .{ "std.BoundedArray", "stdx.BoundedArrayType" },
+        .{ "StaticBitSet", "stdx.BitSetType" },
+        .{ "std.time.Duration", "stdx.Duration" },
+        .{ "std.time.Instant", "stdx.Instant" },
+        .{ "hasUniqueRepresentation", "stdx.has_unique_representation" },
+        .{ "@memcpy(", "stdx.copy_disjoint" },
+        .{ "mem.copyForwards(", "stdx.copy_left" },
+        .{ "mem.copyBackwards(", "stdx.copy_right" },
+        .{ "uintLessThan", "stdx.PRNG" },
+        .{ "intRangeLessThan", "stdx.PRNG" },
+        .{ "intRangeAtMost", "stdx.PRNG" },
+        .{ "intRangeAtMostBiased", "stdx.PRNG" },
+
+        // Library footguns:
+        .{ "unexpectedErrno", "stdx.unexpected_errno" },
+        .{ "posix.send(", "posix.sendto to avoid connection race condition" },
+
+        // Language footguns:
+        .{ "== error.", "switch to avoid silent anyerror upcast" },
+        .{ "!= error.", "switch to avoid silent anyerror upcast" },
+
+        // Everything else:
+        .{ "debug.assert(", "unqualified assert" },
+        .{ "Self = @This()", "proper type name" },
+        .{ "!comptime", "! inside comptime" },
+        .{ "usingnamespace", "something else" },
+    };
+
+    for (ban_list) |ban_item| {
+        const banned, const replacement = ban_item;
+        if (std.mem.indexOf(u8, file.text, banned)) |offset| {
+            errors.add_banned(file, offset, banned, replacement);
+        }
     }
 
-    if (std.mem.indexOf(u8, source, "posix." ++ "send(") != null) {
-        return "use posix.sendto() to avoid connection race condition";
+    // Reminders:
+    // Do use FIXME comments proactively while iterating on the code when you want to make sure
+    // something is revisited before getting into the main branch.
+    inline for (.{ "FIXME", "dbg(" }) |banned| {
+        if (std.mem.indexOf(u8, file.text, banned)) |offset| {
+            if (std.mem.startsWith(u8, file.text[offset..], "dbg(prefix: []const u8")) {
+                // Allow fn dbg( function definition.
+
+            } else {
+                errors.add_banned_reminder(file, offset, banned);
+            }
+        }
     }
+}
 
-    return null;
+test tidy_banned {
+    try check_tidy_file(
+        \\banned.zig
```

Comment:
Use `\\banned.zig` rather than `"banned.zig"` for file name here and elsewhere simply for nicer layout